### PR TITLE
Support for multiple pce's (rebased)

### PR DESCRIPTION
**Pull request by rampxxxx**
Signed-off-by: Javier Garcia <javier.garcia@voltanet.io>

### CODE REVIEW

File: pathd/path_pcep.h [hunk 1]
```diff
@@ -84,6 +85,9 @@ struct pce_opts {
 	struct ipaddr addr;
 	short port;
 	bool draft07;
+	uint8_t precedence;
+	bool is_best;
```
Comment (sylane) [inline pathd/path_pcep.h:89]: This doesn't look like it belong here, these are not options.

File: pathd/path_pcep_controller.c [hunk 2]
```diff
@@ -194,6 +201,7 @@ int pcep_ctrl_initialize(struct thread_master *main_thread,
 	ctrl_state->pcc_count = 0;
 	ctrl_state->pcc_opts =
 		XCALLOC(MTYPE_PCEP, sizeof(*ctrl_state->pcc_opts));
+	memset(ctrl_state->pcc, 0, sizeof(ctrl_state->pcc[0]) * MAX_PCC);
```
Comment (sylane) [inline pathd/path_pcep_controller.c:204]: The memory allocated with XCALLOC is already cleared so there is no need to do any memeset.

Comment (rampxxxx) [reply pathd/path_pcep_controller.c:206]: ok, you're right

File: pathd/path_pcep.h (hunk 1 above)
Comment (rampxxxx) [reply pathd/path_pcep.h:89]: The precedence is a config option and the is_best and previous_best are 'calculated' but if not a new struct must be created and mantained so for now this is a good place

File: pathd/path_pcep_controller.c [hunk 3]
```diff
@@ -377,7 +405,7 @@ int pcep_thread_get_counters_callback(struct thread *t)
 	assert(ctrl_state != NULL);
 	struct pcc_state *pcc_state;
 
-	if (args->pcc_id <= ctrl_state->pcc_count) {
+	if (args->pcc_id <= MAX_PCC) {
```
Comment (sylane) [inline pathd/path_pcep_controller.c:408]: There shouldn't be any more callback using an integer index to do anything.

File: pathd/path_pcep_controller.c [hunk 4]
```diff
@@ -396,9 +424,12 @@ int pcep_thread_send_report_callback(struct thread *t)
 	struct pcc_state *pcc_state;
 
 	if (args->pcc_id == 0) {
```
Comment (sylane) [inline pathd/path_pcep_controller.c:426]: Same hear, I am confused, why do you still use index ?

File: pathd/path_pcep_controller.c [hunk 5]
```diff
@@ -454,13 +485,21 @@ int pcep_thread_timer_handler(struct thread *thread)
 
 	int ret = 0;
 	struct pcc_state *pcc_state = NULL;
-
+	pcc_state = get_pcc_state(ctrl_state, pcc_id);
```
Comment (sylane) [inline pathd/path_pcep_controller.c:488]: Shouldn't be an integer index.

File: pathd/path_pcep_controller.c [hunk 6]
```diff
@@ -454,13 +485,21 @@ int pcep_thread_timer_handler(struct thread *thread)
...
+	assert(pcc_state != NULL);
 	switch (type) {
 	case TM_RECONNECT_PCC:
+		if (pcc_state->retry_count == 1) {
```
Comment (sylane) [inline pathd/path_pcep_controller.c:492]: You shouldn't use the pcc state here, it should be opaque, AFAIK the retry count was given by the pcc when calling the re-connection scheduling function.

File: pathd/path_pcep_controller.c [hunk 7]
```diff
@@ -479,12 +518,15 @@ int pcep_thread_pcep_event(struct thread *thread)
...
 	int i;
 
-	for (i = 0; i < ctrl_state->pcc_count; i++) {
-		struct pcc_state *pcc_state = ctrl_state->pcc[i];
-		if (pcc_state->sess != event->session)
-			continue;
-		pcep_pcc_pcep_event_handler(ctrl_state, pcc_state, event);
-		break;
+	for (i = 0; i < MAX_PCC; i++) {
+		if (ctrl_state->pcc[i]) {
```
Comment (sylane) [inline pathd/path_pcep_controller.c:522]: Shouldn't be using integer indexing anymore.

File: pathd/path_pcep_controller.c [hunk 8]
```diff
@@ -612,8 +658,30 @@ int pcep_thread_event_handler(struct thread *thread)
...
+					get_pcc_state(ctrl_state, pcc_id);
+				pcep_thread_schedule_sync_best_pce(
+					ctrl_state, pcc_id, 30 /*delay*/,
+					&pcc_state->t_reconnect);
+			}
+		}
+		break;
 	case EV_REMOVE_PCC:
-		ret = pcep_thread_event_remove_pcc(ctrl_state, pcc_id);
+		pce_addr = (struct ipaddr *)payload;
```
Comment (sylane) [inline pathd/path_pcep_controller.c:678]: Here we should be sure we own the address, and free it.

File: pathd/path_pcep_controller.c [hunk 9]
```diff
@@ -650,18 +727,62 @@ int pcep_thread_event_update_pcc_options(struct ctrl_state *ctrl_state,
...
+	struct pce_opts *pce_opts_copy;
+	int current_pcc_id = get_pcc_id_by_ip(pcep_g->fpt, &pce_opts->addr);
+	if (!current_pcc_id) {
+		current_pcc_id = pcep_ctrl_get_free_pcc_id(pcep_g->fpt);
+	}
+	if (pcep_g->pce_opts[current_pcc_id - 1] != NULL)
+		XFREE(MTYPE_PCEP, pcep_g->pce_opts[0]);
+	pce_opts_copy = XCALLOC(MTYPE_PCEP, sizeof(*pce_opts));
+	pce_opts_copy = memcpy(pce_opts_copy, pce_opts, sizeof(*pce_opts));
+	pcep_g->pce_opts[current_pcc_id - 1] = pce_opts_copy;
```
Comment (sylane) [inline pathd/path_pcep_controller.c:774]: You should not do that from the controller thread and read it from the main thread without a mutex. This is why this was done in pathd_pcep.c.
In general, pcep_g should only used from the main thread.

File: pathd/path_pcep_controller.c [hunk 10]
```diff
@@ -756,7 +920,162 @@ int pcep_main_event_handler(struct thread *thread)
 
 
 /* ------------ Helper functions ------------ */
+bool is_best_pce(struct ctrl_state *ctrl_state, int pce)
+{
+	if (ctrl_state && ctrl_state->pcc[pce]) {
```
Comment (sylane) [inline pathd/path_pcep_controller.c:925]: This shouldn't be stored there. Two variable in the controller struct that point to the best and maybe the previous best would work fine AFAIK.

File: pathd/path_pcep_controller.c [hunk 11]
```diff
@@ -756,7 +920,162 @@ int pcep_main_event_handler(struct thread *thread)
...
+	int step_0_best = -1;
+	int step_0_previous = -1;
+
+	if (!ctrl_state || !ctrl_state->pcc_count)
+		return 0;
+
+	struct pcc_state **pcc = &ctrl_state->pcc[0];
+
+	// Get state
+	for (int i = 0; i < MAX_PCC; i++) {
```
Comment (sylane) [inline pathd/path_pcep_controller.c:963]: Should be changed for a map...

File: pathd/path_pcep_pcc.c [hunk 12]
```diff
@@ -766,16 +773,17 @@ void specialize_output_path(struct pcc_state *pcc_state, struct path *path)
...
 		is_delegated = path->type == SRTE_CANDIDATE_TYPE_DYNAMIC;
 		/* it seems the PCE consider updating an LSP a creation ?!?
 		at least Cisco does... */
 		was_created = path->update_origin == SRTE_ORIGIN_PCEP;
 	}
 
 	path->pcc_id = pcc_state->id;
-	path->go_active = is_delegated;
-	path->is_delegated = is_delegated;
+	path->go_active = (is_delegated & pcc_state->pce_opts->is_best);
```
Comment (sylane) [inline pathd/path_pcep_pcc.c:785]: This shouldn't com form there, this should be a field of pcc_state structure.

File: pathd/path_pcep_controller.c (hunk 9 above)
Comment (rampxxxx) [reply pathd/path_pcep_controller.c]: Yes, back to pathd_pcep.c as originally was

File: pathd/path_pcep_pcc.c (hunk 12 above)
Comment (rampxxxx) [reply pathd/path_pcep_pcc.c]: ok, is_best and previous_best moved to pcc_state , not in pce_opts anymore

File: pathd/path_pcep_controller.c (hunk 8 above)
Comment (rampxxxx) [reply pathd/path_pcep_controller.c:679]: yes, done

File: pathd/path_pcep_controller.c (hunk 3 above)
Comment (rampxxxx) [reply pathd/path_pcep_controller.c]: I'll keep and array to hold data and the id now is also and int but it's a sequential one no related with array index

File: pathd/path_pcep.c [hunk 13]
```diff
@@ -511,12 +512,34 @@ DEFUN(pcep_cli_no_pce, pcep_cli_no_pce_cmd,
       "Remote PCE server port value\n")
 {
 	/* TODO: Add support for multiple PCE */
+	int i = 2;
+	static struct ipaddr pce_addr;
+	pce_addr.ipa_type = IPADDR_V4;
```
[the remark this answers is not in the document]
Comment (rampxxxx) [reply pathd/path_pcep.c:539]: yes, changed

File: pathd/path_pcep_controller.c [hunk 14]
```diff
@@ -233,11 +242,19 @@ int pcep_ctrl_update_pce_options(struct frr_pthread *fpt, int pcc_id,
...
+int pcep_ctrl_update_timer_best_pce(struct ctrl_state *ctrl_state, int pcc_id)
+{
+	return send_to_thread(ctrl_state, pcc_id, EV_UPDATE_TIMER_BEST_PCE, 0,
+			      NULL);
+}
+int pcep_ctrl_remove_pcc(struct frr_pthread *fpt, struct ipaddr *pce_addr)
 {
 	struct ctrl_state *ctrl_state = get_ctrl_state(fpt);
-	return send_to_thread(ctrl_state, pcc_id, EV_REMOVE_PCC, 0, NULL);
+	return send_to_thread(ctrl_state, 0, EV_REMOVE_PCC, 0, pce_addr);
```
[the remark this answers is not in the document]
Comment (rampxxxx) [reply pathd/path_pcep_controller.c:257]: ok, changed to XCALLOC and XFREE

File: pathd/path_pcep.c [hunk 15]
```diff
@@ -511,12 +512,34 @@ DEFUN(pcep_cli_no_pce, pcep_cli_no_pce_cmd,
...
+	SET_IPADDR_V4(&pce_addr);
+	if (strcmp("ipv6", argv[i]->arg) == 0) {
+		SET_IPADDR_V6(&pce_addr);
+	} else if (strcmp("ip", argv[i]->arg) != 0) {
+		return CMD_ERR_NO_MATCH;
+	}
 
-	pcep_ctrl_remove_pcc(pcep_g->fpt, 1);
-	if (pcep_g->pce_opts[0] != NULL) {
```
[the remark this answers is not in the document]
Comment (rampxxxx) [reply pathd/path_pcep.c:516]: ok, pcep_g handle is back to path_pcep.c

File: pathd/path_pcep_controller.c [hunk 16]
```diff
@@ -400,11 +428,12 @@ int pcep_thread_get_counters_callback(struct thread *t)
 	assert(ctrl_state != NULL);
 	struct pcc_state *pcc_state;
 
-	if (args->pcc_id <= ctrl_state->pcc_count) {
 		pcc_state = get_pcc_state(ctrl_state, args->pcc_id);
```
Comment (sylane) [inline pathd/path_pcep_controller.c:431]: Indentation got wrong ?

File: pathd/path_pcep_controller.c [hunk 17]
```diff
@@ -635,8 +683,30 @@ int pcep_thread_event_handler(struct thread *thread)
...
+			}
+		}
+		break;
 	case EV_REMOVE_PCC:
-		ret = pcep_thread_event_remove_pcc(ctrl_state, pcc_id);
+		pce_opts = (struct pce_opts *)payload;
+		ret = pcep_thread_event_remove_pcc(ctrl_state, pce_opts);
+		new_best_pcc_id = calculate_best_pce(ctrl_state);
+		if (new_best_pcc_id) {
+			ret = pcep_thread_event_update_best_pce(
```
Comment (sylane) [inline pathd/path_pcep_controller.c:707]: This probably dosen't need an event either. And if there is multiple places doing the calculate and notify pccs, it should be done in its own function.

File: pathd/path_pcep_controller.c [hunk 18]
```diff
@@ -673,20 +752,54 @@ int pcep_thread_event_update_pcc_options(struct ctrl_state *ctrl_state,
...
 	return 0;
 }
-
+int pcep_thread_event_update_best_pce(struct ctrl_state *ctrl_state, int best)
+{
+	PCEP_DEBUG(" recalculating pce precedence ");
+	if (best) {
+		struct pcc_state *best_pcc_state =
+			get_pcc_state(ctrl_state, best);
+		if (best_pcc_state->previous_best != best_pcc_state->is_best) {
```
Comment (sylane) [inline pathd/path_pcep_controller.c:761]: Why not have the id of the best pcc and the id of the previous best in the controller state ?

File: pathd/path_pcep_controller.c [hunk 19]
```diff
@@ -673,20 +752,54 @@ int pcep_thread_event_update_pcc_options(struct ctrl_state *ctrl_state,
...
+			PCEP_DEBUG(" %s Resynchro best (%i) previous best (%i)",
+				   best_pcc_state->tag, best_pcc_state->id,
+				   best_pcc_state->previous_best);
+			best_pcc_state->status = PCEP_PCC_SYNCHRONIZING;
```
Comment (sylane) [inline pathd/path_pcep_controller.c:765]: We shouldn't be reading pcc state structure fields, but we should DEFINITELY NOT be modifying them. if there is a good reason, then add a function in the pcc to do what makes sens to do there.

File: pathd/path_pcep_controller.c [hunk 20]
```diff
@@ -673,20 +752,54 @@ int pcep_thread_event_update_pcc_options(struct ctrl_state *ctrl_state,
...
-		set_pcc_state(ctrl_state, pcc_state);
+	int current_pcc_id = get_pcc_id_by_ip_port(ctrl_state, pce_opts);
+	if (current_pcc_id) {
+		pcc_state = get_pcc_state(ctrl_state, current_pcc_id);
 	} else {
-		pcc_state = get_pcc_state(ctrl_state, pcc_id);
+		pcc_state = pcep_pcc_initialize(ctrl_state,
+						get_next_id(ctrl_state));
+		if (set_pcc_state(ctrl_state, pcc_state)) {
+			memset(pcc_state, 0, sizeof(*pcc_state));
```
Comment (sylane) [inline pathd/path_pcep_controller.c:796]: Why are you doing this ?

File: pathd/path_pcep_controller.c [hunk 21]
```diff
@@ -673,20 +752,54 @@ int pcep_thread_event_update_pcc_options(struct ctrl_state *ctrl_state,
...
+			XFREE(MTYPE_PCEP, pcc_state);
+			return 0;
+		}
 	}
 
+
```
Comment (sylane) [inline pathd/path_pcep_controller.c:802]: extra line

File: pathd/path_pcep_controller.c [hunk 22]
```diff
@@ -696,17 +809,51 @@ int pcep_thread_event_update_pce_options(struct ctrl_state *ctrl_state,
...
+		if (pcc_state) {
+			remove_pcc_state(ctrl_state, pcc_state);
+			pcep_pcc_finalize(ctrl_state, pcc_state);
+		}
+	}
+	return 0;
+}
+int pcep_thread_event_remove_pcc_all(struct ctrl_state *ctrl_state)
+{
+	if (!ctrl_state) {
```
Comment (sylane) [inline pathd/path_pcep_controller.c:830]: Is there a sane recoverable situations where the controller state could be NULL ? If not this should be an assert.

File: pathd/path_pcep_controller.c [hunk 23]
```diff
@@ -696,17 +809,51 @@ int pcep_thread_event_update_pce_options(struct ctrl_state *ctrl_state,
...
+	for (int i = 0; i < MAX_PCC; i++) {
+		pcep_thread_event_remove_pcc_by_id(
+			ctrl_state, get_pcc_id_by_idx(ctrl_state, i));
+	}
+	return 0;
+}
+int pcep_thread_event_remove_pcc(struct ctrl_state *ctrl_state,
+				 struct pce_opts *pce_opts)
+{
+	if (!ctrl_state) {
```
Comment (sylane) [inline pathd/path_pcep_controller.c:843]: Same here, could this function be called with ctrl_state being NULL ?

File: pathd/path_pcep_controller.c (hunk 11 above)
Comment (sylane) [reply pathd/path_pcep_controller.c:975]: The problem is not the memory usage or the efficiency, it is about the complexity of having to deal with multiple types of indexes. We can go with that for now, there is more important changes, but this should be changed to a tre or a hash map soon.

File: pathd/path_pcep_controller.c [hunk 24]
```diff
@@ -779,7 +929,161 @@ int pcep_main_event_handler(struct thread *thread)
...
+					}
+				}
+			}
+		}
+	}
+
+
+	return best_pce == -1 ? 0 : pcc[best_pce]->id;
+}
+int pcep_ctrl_get_free_pcc_idx(struct pcc_state **pcc_state)
```
Comment (sylane) [inline pathd/path_pcep_controller.c:1072]: Why not get the controller state as parameter, this doesn't fell safe to assume the size of what you get as a parameter is MAX_PCC, you can only assume that from the array in the controller state.

File: pathd/path_pcep_controller.h [hunk 25]
```diff
@@ -75,22 +80,35 @@ struct pcep_ctrl_socket_data {
 typedef int (*pcep_ctrl_thread_callback)(struct thread *);
 
 /* Functions called from the main thread */
+int calculate_best_pce(struct ctrl_state *ctrl_state);
```
Comment (sylane) [inline pathd/path_pcep_controller.h:83]: All these functions are called only from the main thread ? if it is the case they should be names `pcep_ctrl_XXX`.
But from what I can see they are not and the prototype should be moved to the corresponding section.
Are all these functions used outside of `path_pcep_controller.c` ? if not the SHOULD NOT be there.

File: pathd/path_pcep_pcc.c [hunk 26]
```diff
@@ -542,9 +547,11 @@ void pcep_pcc_pcep_event_handler(struct ctrl_state *ctrl_state,
 		PCEP_DEBUG_PCEP("%s PCEP message: %s", pcc_state->tag,
 				format_pcep_message(event->message));
 		break;
+	case PCE_DEAD_TIMER_EXPIRED:
+		PCEP_DEBUG("%s PCE_DEAD_TIMER_EXPIRED %d", pcc_state->tag,
```
Comment (sylane) [inline pathd/path_pcep_pcc.c:551]: The received even is already logged before, this is only to have the extra retry counter. I am not sure we need that extra logging. I assume you added it for debug and it wasn't cleaned up.

File: pathd/path_pcep_cli.c [hunk 27]
```diff
@@ -856,7 +857,14 @@ static int path_pcep_cli_pcc_delete(struct vty *vty, const char *ip_str,
 				    struct in6_addr *ipv6, const char *port_str,
 				    long port, const char *msd_str, long msd)
 {
-	pcep_ctrl_remove_pcc(pcep_g->fpt, 1);
+	pcep_ctrl_remove_pcc(pcep_g->fpt, NULL);
+
+	for (int i = 0; i < MAX_PCE; i++) {
+		if (pcep_g->pce_opts_cli[i] != NULL) {
+			XFREE(MTYPE_PCEP, pcep_g->pce_opts_cli[i]);
+			pcep_g->pcc_opts = NULL;
```
[the remark this answers is not in the document]
Comment (rampxxxx) [reply pathd/path_pcep_cli.c:868]: thank you, yes, a prob with the rebase

File: pathd/path_pcep_cli.c [hunk 28]
```diff
@@ -936,7 +937,9 @@ static int path_pcep_cli_pcc_pcc_peer_delete(struct vty *vty,
 
 	/* TODO when Multi-PCE is added, the PCC Id needs to be set
```
[the remark this answers is not in the document]
Comment (rampxxxx) [reply pathd/path_pcep_cli.c:941]: Deleted, as has no sense now.

File: pathd/path_pcep_controller.c (hunk 20 above)
Comment (rampxxxx) [reply pathd/path_pcep_controller.c:796]: For  past experiences about reusing mem without initializtion but as far as we use heare XCALLOC it's not needed.

File: pathd/path_pcep_pcc.c (hunk 26 above)
Comment (rampxxxx) [reply pathd/path_pcep_pcc.c:551]: yes, deleted

File: pathd/path_pcep_controller.c (hunk 16 above)
Comment (rampxxxx) [reply pathd/path_pcep_controller.c:431]: changed

File: pathd/path_pcep_controller.c (hunk 22 above)
Comment (rampxxxx) [reply pathd/path_pcep_controller.c:828]: done

File: pathd/path_pcep_controller.c (hunk 23 above)
Comment (rampxxxx) [reply pathd/path_pcep_controller.c:841]: done

File: pathd/path_pcep_controller.c (hunk 24 above)
Comment (rampxxxx) [reply pathd/path_pcep_controller.c:1070]: changed

File: pathd/path_pcep_controller.c [hunk 29]
```diff
@@ -330,6 +347,15 @@ void pcep_thread_update_path(struct ctrl_state *ctrl_state, int pcc_id,
 		     path);
 }
 
+void pcep_thread_schedule_sync_best_pce(struct ctrl_state *ctrl_state,
+					int pcc_id, int delay,
+					struct thread **thread)
+{
+
```
[the remark this answers is not in the document]
Comment (rampxxxx) [reply pathd/path_pcep_controller.c:351]: done

File: pathd/path_pcep_cli.c (hunk 27 above)
[the remark this answers is not in the document]
Comment (rampxxxx) [reply pathd/path_pcep_cli.c]: fixed

File: pathd/path_pcep_controller.h (hunk 25 above)
Comment (rampxxxx) [reply pathd/path_pcep_controller.h]: changed

File: pathd/path_pcep_controller.c [hunk 30]
```diff
@@ -226,17 +236,24 @@ int pcep_ctrl_update_pcc_options(struct frr_pthread *fpt, struct pcc_opts *opts)
...
-				 struct pce_opts *opts)
+int pcep_ctrl_update_pce_options(struct frr_pthread *fpt, struct pce_opts *opts)
 {
 	struct ctrl_state *ctrl_state = get_ctrl_state(fpt);
-	return send_to_thread(ctrl_state, pcc_id, EV_UPDATE_PCE_OPTS, 0, opts);
+	return send_to_thread(ctrl_state, 0, EV_UPDATE_PCE_OPTS, 0, opts);
 }
-
-int pcep_ctrl_remove_pcc(struct frr_pthread *fpt, int pcc_id)
+int pcep_ctrl_update_best_pce(struct ctrl_state *ctrl_state, int pcc_id)
```
[the remark this answers is not in the document]
Comment (rampxxxx) [reply pathd/path_pcep_controller.c]: changed

File: pathd/path_pcep_controller.c (hunk 21 above)
Comment (rampxxxx) [reply pathd/path_pcep_controller.c]: deleted

File: pathd/path_pcep_controller.c (hunk 19 above)
Comment (rampxxxx) [reply pathd/path_pcep_controller.c]: moved to a function

File: pathd/path_pcep_cli.c [hunk 31]
```diff
@@ -1036,6 +1031,12 @@ int pcep_cli_pcc_config_write(struct vty *vty)
 			continue;
 		}
 
+		/* Only show connected PCEs */
```
Comment (sylane) [inline pathd/path_pcep_cli.c]: Why is it filtering some PCE out ?

File: pathd/path_pcep_controller.c (hunk 16 above)
Comment (sylane) [reply pathd/path_pcep_controller.c:426]: The indentation still looks wrong, at least here in the github review interface.

File: pathd/path_pcep_pcc.c [hunk 32]
```diff
@@ -466,6 +477,13 @@ void pcep_pcc_send_report(struct ctrl_state *ctrl_state,
 	}
 }
 
+void pcep_pcc_start_sync(struct ctrl_state *ctrl_state, struct pcc_state *pcc_state)
```
Comment (sylane) [inline pathd/path_pcep_pcc.c:480]: Synchronization cannot be done anytime during the PCEP connection, only as the first thing. When changing PCE, we should send all the reports but without the synchronization flag and we should NOT send the last NULL LSP.

File: pathd/path_pcep_pcc.c [hunk 33]
```diff
@@ -809,6 +827,11 @@ void schedule_reconnect(struct ctrl_state *ctrl_state,
 	pcep_thread_schedule_reconnect(ctrl_state, pcc_state->id,
 				       pcc_state->retry_count,
 				       &pcc_state->t_reconnect);
+  if(pcc_state->retry_count==1){
```
Comment (sylane) [inline pathd/path_pcep_pcc.c:830]: Should be `if (..) {` with spaces

File: pathd/path_pcep_pcc.c [hunk 34]
```diff
@@ -868,8 +891,8 @@ void specialize_outgoing_path(struct pcc_state *pcc_state, struct path *path)
 	was_created = path->update_origin == SRTE_ORIGIN_PCEP;
 
 	path->pcc_id = pcc_state->id;
-	path->go_active = is_delegated;
-	path->is_delegated = is_delegated;
+	path->go_active = (is_delegated & pcc_state->is_best);
```
Comment (sylane) [inline pathd/path_pcep_pcc.c:894]: Should be `is_delegated && pcc_state->is_best` with `&` and without parenthesis.

File: pathd/path_pcep_pcc.c [hunk 35]
```diff
@@ -868,8 +891,8 @@ void specialize_outgoing_path(struct pcc_state *pcc_state, struct path *path)
...
+	path->is_delegated = (is_delegated & pcc_state->is_best);
```
Comment (sylane) [inline pathd/path_pcep_pcc.c:895]: Same here

File: pathd/path_pcep_pcc.c (hunk 34 above)
Comment (rampxxxx) [reply pathd/path_pcep_pcc.c:894]: done

File: pathd/path_pcep_pcc.c (hunk 35 above)
Comment (rampxxxx) [reply pathd/path_pcep_pcc.c:895]: done

File: pathd/path_pcep_cli.c (hunk 31 above)
Comment (rampxxxx) [reply pathd/path_pcep_cli.c]: done

**Pull request merged**